File: ex509.java
class Solution {
    private int[] dp;
    
    private int recur(int N) {
        if (dp[N] != -1)    return dp[N];
        
        int a = recur(N - 1) + recur(N - 2);
        dp[N] = a;
        
        return dp[N];
    }
    
    public int fib(int N) {
        if (N == 0)    return 0;
        
        dp = new int[N + 1];
        Arrays.fill(dp, -1);
        dp[0] = 0;    dp[1] = 1;
        
        return recur(N);
    }
}

File: code/ex509.cpp
class Solution {
public:
    int fib(int n) {
        if (n <= 1)    return n;
        int p1 = 0, p = 1;
        for (int i = 2; i <= n; i++) {
            int newP = p1 + p;
            p1 = p;
            p = newP;
        }
        return p;
    }
};